### ISSUE TITLE
Special Tokens

### ISSUE THREAD

**Issue by anthonyfuller7** [aviclu/CD-LM#3]
Hi there,

Thanks for posting your code and uploading the pretrained CD-LM model!

I've read through the readme, and code - but wanted to get some clarification. When using CD-LM for simple MLM, I am doing this:
- Wrapping each sentence in <s> </s> tags
- Wrapping each document in <doc-s> </doc-s> tags
- Randomly assigning masks with p=15%
- Making sure the special tags (above) are not masked
- Assigning global attention (1s) to all masked positions, and 0s elsewhere
- Assigning label=-100 to all non-masked positions
- Concatenating all document ids, global attention masks, and labels for that sample

Is this correct?

```
def prepare_cd_lm(document_text, prob):
  sentences = sent_tokenize(document_text)
  sentence_ids_list = []
  global_att_list = []
  labels_list = []

  for sen in sentences:
    sen_ids = tokenizer.encode(sen, return_tensors='pt').view(-1)
    probability_matrix = torch.full(sen_ids.shape, prob)
    masked_indices = torch.bernoulli(probability_matrix).bool()

    labels = sen_ids.clone()
    labels[~masked_indices] = -100
    global_attention_mask = torch.zeros_like(labels)
    global_attention_mask[masked_indices] = 1
    sen_ids[masked_indices] = 50264
    sen_ids[0] = 0
    sen_ids[-1] = 2
    global_attention_mask[0] = 0
    global_attention_mask[-1] = 0
    labels[0] = -100
    labels[-1] = -100

    sentence_ids_list.append(sen_ids)
    global_att_list.append(global_attention_mask)
    labels_list.append(labels)
  
  sentence_ids_list = torch.cat(sentence_ids_list).view(-1)
  global_att_list = torch.cat(global_att_list).view(-1)
  labels_list = torch.cat(labels_list).view(-1)

  start_doc_id = torch.LongTensor([50265])
  end_doc_id = torch.LongTensor([50266])

  extra_label = torch.LongTensor([-100])
  extra_att = torch.LongTensor([0])

  sentence_ids_list = torch.cat([start_doc_id, sentence_ids_list, end_doc_id], dim=0).view(-1)
  global_att_list = torch.cat([extra_att, global_att_list, extra_att], dim=0).view(-1)
  labels_list = torch.cat([extra_label, labels_list, extra_label], dim=0).view(-1)

  return sentence_ids_list, global_att_list, labels_list
```

```
losses = []
count = 0
for idx in range(len(dataset['test'])):
  docs = dataset['test'][idx]['document']
  docs = docs.split('|||||')[:-1]

  if len(docs) >= 3:
    docs = docs[:3]
    all_ids = []
    all_global_att = []
    all_labels = []
    for i, input_text in enumerate(docs):
      doc_ids, doc_global_att, doc_labels = prepare_cd_lm(document_text=input_text, prob=0.15)
      all_ids.append(doc_ids)
      all_global_att.append(doc_global_att)
      all_labels.append(doc_labels)

    all_ids = torch.cat(all_ids, dim=-1).view(1, -1)
    all_global_att = torch.cat(all_global_att, dim=-1).view(1, -1)
    all_labels = torch.cat(all_labels, dim=-1).view(1, -1)
    count += 1
    with torch.no_grad():
      loss = model(input_ids=all_ids.cuda(), global_attention_mask=all_global_att.cuda(), labels=all_labels.cuda(), return_dict=True)['loss'].item()
    losses.append(loss)

print(f'Loss: {torch.Tensor(losses).mean()} Total: {count}')
```

**Comment by anthonyfuller7** [comment]
bug in code

**Comment by anthonyfuller7** [comment]
When I first run the pretrained CDLM model, with the above tokenization scheme, I get very high loss. But over the course of just a couple hundreds training steps, the loss rapidly decreases. This leads me to believe I had something wrong, but the model very quickly adjusted.
